Test hangs forever with a specific exception chain

When running a test that raises an Exception with a loop in the `__causes__` variable (by using `raise A from B`), pytest hangs forever but a normal execution succeeds.

This is a minimal example I have created to reproduce this issue:

```python
class SomethingWentWrong(Exception):
    pass


def run_safely(callable):
    """
    Runs the callable and returns it,
    or raises SomethingWentWrong if there is any exception
    """
    try:
        return callable()
    except Exception as e:
        raise SomethingWentWrong() from e


def do_error():
    return 0 / 0


def do_something():
    try:
        return run_safely(lambda: do_error())
    except SomethingWentWrong as exc:
        raise exc.__cause__


if __name__ == "__main__":
    do_something()


def test_bug():
    do_something()
```

Note inside the `except` clause in `do_something()` it is reraising the original clause of the exception, and my guess is that implicitly it is creating a loop there.

A normal execution runs fine and shows the chained exception:

```
$ python test_bug.py
Traceback (most recent call last):
  File "test_bug.py", line 22, in do_something
    return run_safely(lambda: do_error())
  File "test_bug.py", line 13, in run_safely
    raise SomethingWentWrong() from e
__main__.SomethingWentWrong

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_bug.py", line 30, in <module>
    do_something()
  File "test_bug.py", line 26, in do_something
    raise exc.__cause__
  File "test_bug.py", line 11, in run_safely
    return callable()
  File "test_bug.py", line 22, in <lambda>
    return run_safely(lambda: do_error())
  File "test_bug.py", line 17, in do_error
    return 0 / 0
ZeroDivisionError: division by zero
```

But when running with pytest it just hangs forever:
```
$ pytest test_bug.py
=============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.6.6, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /Users/nova/workspace/sherlock, inifile:
collected 1 item

test_bug.py
```

Version information:
``` 
$ python -V
Python 3.6.6
$ pip list
Package        Version
-------------- -------
atomicwrites   1.1.5
attrs          18.1.0
more-itertools 4.3.0
pip            18.0
pluggy         0.7.1
py             1.5.4
pytest         3.7.1
setuptools     40.0.0
six            1.11.0
$ uname -a
Darwin 10.0.0.10 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
```

confirmed, goes away with `--tb=native`

Here's an interrupted run with `--fulltrace`:

```
$ pytest t.py --tb=short --fulltrace
============================= test session starts ==============================
platform linux -- Python 3.6.5, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py ^C

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

config = <_pytest.config.Config object at 0x7f3f977266a0>
doit = <function _main at 0x7f3f97bb5400>

    def wrap_session(config, doit):
        """Skeleton command line program"""
        session = Session(config)
        session.exitstatus = EXIT_OK
        initstate = 0
        try:
            try:
                config._do_configure()
                initstate = 1
                config.hook.pytest_sessionstart(session=session)
                initstate = 2
>               session.exitstatus = doit(config, session) or 0

venv/lib/python3.6/site-packages/_pytest/main.py:178: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

config = <_pytest.config.Config object at 0x7f3f977266a0>
session = <Session 't'>

    def _main(config, session):
        """ default command line protocol for initialization, session,
        running tests and reporting. """
        config.hook.pytest_collection(session=session)
>       config.hook.pytest_runtestloop(session=session)

venv/lib/python3.6/site-packages/_pytest/main.py:215: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtestloop'>, args = ()
kwargs = {'session': <Session 't'>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.argnames:
            notincall = set(self.argnames) - set(['__multicall__']) - set(
                kwargs.keys())
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call"
                    .format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)

venv/lib/python3.6/site-packages/pluggy/hooks.py:258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x7f3f9b53f358>
hook = <_HookCaller 'pytest_runtestloop'>
methods = [<HookImpl plugin_name='main', plugin=<module '_pytest.main' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/main.py'>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7f3f97163f60>>]
kwargs = {'session': <Session 't'>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

venv/lib/python3.6/site-packages/pluggy/manager.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtestloop'>
methods = [<HookImpl plugin_name='main', plugin=<module '_pytest.main' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/main.py'>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7f3f97163f60>>]
kwargs = {'session': <Session 't'>}

    self._inner_hookexec = lambda hook, methods, kwargs: \
        hook.multicall(
            methods, kwargs,
>           firstresult=hook.spec_opts.get('firstresult'),
        )

venv/lib/python3.6/site-packages/pluggy/manager.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <Session 't'>

    def pytest_runtestloop(session):
        if session.testsfailed and not session.config.option.continue_on_collection_errors:
            raise session.Interrupted("%d errors during collection" % session.testsfailed)
    
        if session.config.option.collectonly:
            return True
    
        for i, item in enumerate(session.items):
            nextitem = session.items[i + 1] if i + 1 < len(session.items) else None
>           item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)

venv/lib/python3.6/site-packages/_pytest/main.py:236: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_protocol'>, args = ()
kwargs = {'item': <Function 'test_bug'>, 'nextitem': None}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.argnames:
            notincall = set(self.argnames) - set(['__multicall__']) - set(
                kwargs.keys())
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call"
                    .format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)

venv/lib/python3.6/site-packages/pluggy/hooks.py:258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x7f3f9b53f358>
hook = <_HookCaller 'pytest_runtest_protocol'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest...ame='warnings', plugin=<module '_pytest.warnings' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/warnings.py'>>]
kwargs = {'item': <Function 'test_bug'>, 'nextitem': None}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

venv/lib/python3.6/site-packages/pluggy/manager.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_protocol'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest...ame='warnings', plugin=<module '_pytest.warnings' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/warnings.py'>>]
kwargs = {'item': <Function 'test_bug'>, 'nextitem': None}

    self._inner_hookexec = lambda hook, methods, kwargs: \
        hook.multicall(
            methods, kwargs,
>           firstresult=hook.spec_opts.get('firstresult'),
        )

venv/lib/python3.6/site-packages/pluggy/manager.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function 'test_bug'>, nextitem = None

    def pytest_runtest_protocol(item, nextitem):
        item.ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
>       runtestprotocol(item, nextitem=nextitem)

venv/lib/python3.6/site-packages/_pytest/runner.py:66: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function 'test_bug'>, log = True, nextitem = None

    def runtestprotocol(item, log=True, nextitem=None):
        hasrequest = hasattr(item, "_request")
        if hasrequest and not item._request:
            item._initrequest()
        rep = call_and_report(item, "setup", log)
        reports = [rep]
        if rep.passed:
            if item.config.option.setupshow:
                show_test_item(item)
            if not item.config.option.setuponly:
>               reports.append(call_and_report(item, "call", log))

venv/lib/python3.6/site-packages/_pytest/runner.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function 'test_bug'>, when = 'call', log = True, kwds = {}
call = <CallInfo when='call' exception: division by zero>
hook = <pluggy.hooks._HookRelay object at 0x7f3f9b5363c8>

    def call_and_report(item, when, log=True, **kwds):
        call = call_runtest_hook(item, when, **kwds)
        hook = item.ihook
>       report = hook.pytest_runtest_makereport(item=item, call=call)

venv/lib/python3.6/site-packages/_pytest/runner.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_makereport'>, args = ()
kwargs = {'call': <CallInfo when='call' exception: division by zero>, 'item': <Function 'test_bug'>}
notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.argnames:
            notincall = set(self.argnames) - set(['__multicall__']) - set(
                kwargs.keys())
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call"
                    .format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)

venv/lib/python3.6/site-packages/pluggy/hooks.py:258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x7f3f9b53f358>
hook = <_HookCaller 'pytest_runtest_makereport'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest...ame='skipping', plugin=<module '_pytest.skipping' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/skipping.py'>>]
kwargs = {'call': <CallInfo when='call' exception: division by zero>, 'item': <Function 'test_bug'>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

venv/lib/python3.6/site-packages/pluggy/manager.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_makereport'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest...ame='skipping', plugin=<module '_pytest.skipping' from '/tmp/t/venv/lib/python3.6/site-packages/_pytest/skipping.py'>>]
kwargs = {'call': <CallInfo when='call' exception: division by zero>, 'item': <Function 'test_bug'>}

    self._inner_hookexec = lambda hook, methods, kwargs: \
        hook.multicall(
            methods, kwargs,
>           firstresult=hook.spec_opts.get('firstresult'),
        )

venv/lib/python3.6/site-packages/pluggy/manager.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function 'test_bug'>
call = <CallInfo when='call' exception: division by zero>

    @hookimpl(hookwrapper=True)
    def pytest_runtest_makereport(item, call):
        outcome = yield
>       rep = outcome.get_result()

venv/lib/python3.6/site-packages/_pytest/skipping.py:123: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function 'test_bug'>
call = <CallInfo when='call' exception: division by zero>

    def pytest_runtest_makereport(item, call):
        when = call.when
        duration = call.stop - call.start
        keywords = {x: 1 for x in item.keywords}
        excinfo = call.excinfo
        sections = []
        if not call.excinfo:
            outcome = "passed"
            longrepr = None
        else:
            if not isinstance(excinfo, ExceptionInfo):
                outcome = "failed"
                longrepr = excinfo
            elif excinfo.errisinstance(skip.Exception):
                outcome = "skipped"
                r = excinfo._getreprcrash()
                longrepr = (str(r.path), r.lineno, r.message)
            else:
                outcome = "failed"
                if call.when == "call":
>                   longrepr = item.repr_failure(excinfo)

venv/lib/python3.6/site-packages/_pytest/runner.py:240: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Function 'test_bug'>
excinfo = <ExceptionInfo ZeroDivisionError tblen=22>, outerr = None

    def repr_failure(self, excinfo, outerr=None):
        assert outerr is None, "XXX outerr usage is deprecated"
        style = self.config.option.tbstyle
        if style == "auto":
            style = "long"
>       return self._repr_failure_py(excinfo, style=style)

venv/lib/python3.6/site-packages/_pytest/python.py:754: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Function 'test_bug'>
excinfo = <ExceptionInfo ZeroDivisionError tblen=22>, style = 'short'

    def _repr_failure_py(self, excinfo, style="long"):
        if excinfo.errisinstance(fail.Exception):
            if not excinfo.value.pytrace:
                return py._builtin._totext(excinfo.value)
>       return super(FunctionMixin, self)._repr_failure_py(excinfo, style=style)

venv/lib/python3.6/site-packages/_pytest/python.py:747: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Function 'test_bug'>
excinfo = <ExceptionInfo ZeroDivisionError tblen=22>, style = 'long'

    def _repr_failure_py(self, excinfo, style=None):
        fm = self.session._fixturemanager
        if excinfo.errisinstance(fm.FixtureLookupError):
            return excinfo.value.formatrepr()
        tbfilter = True
        if self.config.option.fulltrace:
            style = "long"
        else:
            tb = _pytest._code.Traceback([excinfo.traceback[-1]])
            self._prunetraceback(excinfo)
            if len(excinfo.traceback) == 0:
                excinfo.traceback = tb
            tbfilter = False  # prunetraceback already does it
            if style == "auto":
                style = "long"
        # XXX should excinfo.getrepr record all data and toterminal() process it?
        if style is None:
            if self.config.option.tbstyle == "short":
                style = "short"
            else:
                style = "long"
    
        if self.config.option.verbose > 1:
            truncate_locals = False
        else:
            truncate_locals = True
    
        try:
            os.getcwd()
            abspath = False
        except OSError:
            abspath = True
    
        return excinfo.getrepr(
            funcargs=True,
            abspath=abspath,
            showlocals=self.config.option.showlocals,
            style=style,
            tbfilter=tbfilter,
>           truncate_locals=truncate_locals,
        )

venv/lib/python3.6/site-packages/_pytest/nodes.py:307: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ExceptionInfo ZeroDivisionError tblen=22>, showlocals = False
style = 'long', abspath = False, tbfilter = True, funcargs = True
truncate_locals = True

    def getrepr(
        self,
        showlocals=False,
        style="long",
        abspath=False,
        tbfilter=True,
        funcargs=False,
        truncate_locals=True,
    ):
        """ return str()able representation of this exception info.
                showlocals: show locals per traceback entry
                style: long|short|no|native traceback style
                tbfilter: hide entries (where __tracebackhide__ is true)
    
                in case of style==native, tbfilter and showlocals is ignored.
            """
        if style == "native":
            return ReprExceptionInfo(
                ReprTracebackNative(
                    traceback.format_exception(
                        self.type, self.value, self.traceback[0]._rawentry
                    )
                ),
                self._getreprcrash(),
            )
    
        fmt = FormattedExcinfo(
            showlocals=showlocals,
            style=style,
            abspath=abspath,
            tbfilter=tbfilter,
            funcargs=funcargs,
            truncate_locals=truncate_locals,
        )
>       return fmt.repr_excinfo(self)

venv/lib/python3.6/site-packages/_pytest/_code/code.py:479: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = FormattedExcinfo(showlocals=False, style='long', abspath=False, tbfilter=True, funcargs=True, truncate_locals=True)
excinfo = <ExceptionInfo ZeroDivisionError tblen=22>

    def repr_excinfo(self, excinfo):
        if _PY2:
            reprtraceback = self.repr_traceback(excinfo)
            reprcrash = excinfo._getreprcrash()
    
            return ReprExceptionInfo(reprtraceback, reprcrash)
        else:
            repr_chain = []
            e = excinfo.value
            descr = None
            while e is not None:
                if excinfo:
>                   reprtraceback = self.repr_traceback(excinfo)

venv/lib/python3.6/site-packages/_pytest/_code/code.py:724: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = FormattedExcinfo(showlocals=False, style='long', abspath=False, tbfilter=True, funcargs=True, truncate_locals=True)
excinfo = <ExceptionInfo ZeroDivisionError tblen=22>

    def repr_traceback(self, excinfo):
        traceback = excinfo.traceback
        if self.tbfilter:
            traceback = traceback.filter()
    
        if is_recursion_error(excinfo):
            traceback, extraline = self._truncate_recursive_traceback(traceback)
        else:
            extraline = None
    
        last = traceback[-1]
        entries = []
        for index, entry in enumerate(traceback):
            einfo = (last == entry) and excinfo or None
>           reprentry = self.repr_traceback_entry(entry, einfo)

venv/lib/python3.6/site-packages/_pytest/_code/code.py:671: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = FormattedExcinfo(showlocals=False, style='long', abspath=False, tbfilter=True, funcargs=True, truncate_locals=True)
entry = <TracebackEntry /tmp/t/venv/lib/python3.6/site-packages/_pytest/python.py:196>
excinfo = None

    def repr_traceback_entry(self, entry, excinfo=None):
        import _pytest._code
    
>       source = self._getentrysource(entry)

venv/lib/python3.6/site-packages/_pytest/_code/code.py:616: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = FormattedExcinfo(showlocals=False, style='long', abspath=False, tbfilter=True, funcargs=True, truncate_locals=True)
entry = <TracebackEntry /tmp/t/venv/lib/python3.6/site-packages/_pytest/python.py:196>

    def _getentrysource(self, entry):
>       source = entry.getsource(self.astcache)

venv/lib/python3.6/site-packages/_pytest/_code/code.py:536: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <TracebackEntry /tmp/t/venv/lib/python3.6/site-packages/_pytest/python.py:196>
astcache = {local('/tmp/t/t.py'): <_ast.Module object at 0x7f3f96e08908>, local('/tmp/t/venv/lib/python3.6/site-packages/_pytest/...f0df28>, local('/tmp/t/venv/lib/python3.6/site-packages/pluggy/hooks.py'): <_ast.Module object at 0x7f3f96e77ba8>, ...}

    def getsource(self, astcache=None):
        """ return failing source code. """
        # we use the passed in astcache to not reparse asttrees
        # within exception info printing
        from _pytest._code.source import getstatementrange_ast
    
        source = self.frame.code.fullsource
        if source is None:
            return None
        key = astnode = None
        if astcache is not None:
            key = self.frame.code.path
            if key is not None:
                astnode = astcache.get(key, None)
        start = self.getfirstlinesource()
        try:
            astnode, _, end = getstatementrange_ast(
>               self.lineno, source, astnode=astnode
            )

venv/lib/python3.6/site-packages/_pytest/_code/code.py:221: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

lineno = 195, source = <_pytest._code.source.Source object at 0x7f3f96d5ada0>
assertion = False, astnode = <_ast.Module object at 0x7f3f96e08c18>

    def getstatementrange_ast(lineno, source, assertion=False, astnode=None):
        if astnode is None:
            content = str(source)
            astnode = compile(content, "source", "exec", 1024)  # 1024 for AST
    
>       start, end = get_statement_startend2(lineno, astnode)

venv/lib/python3.6/site-packages/_pytest/_code/source.py:343: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

lineno = 195, node = <_ast.Module object at 0x7f3f96e08c18>

    def get_statement_startend2(lineno, node):
        import ast
    
        # flatten all statements and except handlers into one lineno-list
        # AST's line numbers start indexing at 1
        values = []
        for x in ast.walk(node):
>           if isinstance(x, (ast.stmt, ast.ExceptHandler)):
E           KeyboardInterrupt

venv/lib/python3.6/site-packages/_pytest/_code/source.py:321: KeyboardInterrupt
========================= no tests ran in 1.56 seconds =========================
```

--- a/changelog/3804.bugfix.rst
+++ b/changelog/3804.bugfix.rst
@@ -1,0 +1,1 @@
+Fix traceback reporting for exceptions with ``__cause__`` cycles.

--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -719,7 +719,9 @@ class FormattedExcinfo(object):
             repr_chain = []
             e = excinfo.value
             descr = None
-            while e is not None:
+            seen = set()
+            while e is not None and id(e) not in seen:
+                seen.add(id(e))
                 if excinfo:
                     reprtraceback = self.repr_traceback(excinfo)
                     reprcrash = excinfo._getreprcrash()

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -4,6 +4,7 @@ from __future__ import absolute_import, division, print_function
 import operator
 import os
 import sys
+import textwrap
 import _pytest
 import py
 import pytest
@@ -1264,6 +1265,50 @@ raise ValueError()
                 "E *RuntimeError: runtime problem",
             ]
         )
+
+    @pytest.mark.skipif("sys.version_info[0] < 3")
+    def test_exc_chain_repr_cycle(self, importasmod):
+        mod = importasmod(
+            """
+            class Err(Exception):
+                pass
+            def fail():
+                return 0 / 0
+            def reraise():
+                try:
+                    fail()
+                except ZeroDivisionError as e:
+                    raise Err() from e
+            def unreraise():
+                try:
+                    reraise()
+                except Err as e:
+                    raise e.__cause__
+        """
+        )
+        excinfo = pytest.raises(ZeroDivisionError, mod.unreraise)
+        r = excinfo.getrepr(style="short")
+        tw = TWMock()
+        r.toterminal(tw)
+        out = "\n".join(line for line in tw.lines if isinstance(line, str))
+        expected_out = textwrap.dedent(
+            """\
+            :13: in unreraise
+                reraise()
+            :10: in reraise
+                raise Err() from e
+            E   test_exc_chain_repr_cycle0.mod.Err
+
+            During handling of the above exception, another exception occurred:
+            :15: in unreraise
+                raise e.__cause__
+            :8: in reraise
+                fail()
+            :5: in fail
+                return 0 / 0
+            E   ZeroDivisionError: division by zero"""
+        )
+        assert out == expected_out
 
 
 @pytest.mark.parametrize("style", ["short", "long"])

